lib/plugins/input/kubernetesEvents.js

```diff
@@ -1,45 +1,55 @@
 // Emit k8s events as Node.js events
+const { KubeConfig } = require('kubernetes-client')
 const Client = require('kubernetes-client').Client
-const config = require('kubernetes-client').config
-const JSONStream = require('json-stream')
-var consoleLogger = require('../../util/logger.js')
-var EventEmitter = require('events').EventEmitter
+const Request = require('kubernetes-client/backends/request')
+const consoleLogger = require('../../util/logger.js')
+const EventEmitter = require('events').EventEmitter
 
 class KubernetesEventEmitter extends EventEmitter {
   constructor () {
     super()
+    const kubeconfig = new KubeConfig()
     // do we run in k8s cluster?
     if (process.env.KUBERNETES_PORT_443_TCP !== undefined) {
-      this.client = new Client({ config: config.getInCluster() })
+      // this.client = new Client({ config: config.getInCluster() })
+      kubeconfig.loadFromCluster()   
     } else {
-      this.client = new Client({ config: config.fromKubeconfig() })
+      kubeconfig.loadFromDefault()
     }
+    this.client = new Client({ backend: new Request({ kubeconfig }) })
     this.client.loadSpec().then(this.start.bind(this)).catch(error => {
-      console.error(error)
+      console.error('Error in k8s client', error)
     })
   }
 
-  startWatching () {
+  async startWatching (namespace, cb) {
     var self = this
     // API docs: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.14/#list-all-namespaces-event-v1-core
-    self.stream = self.client.api.v1.events.getStream()
-    self.jsonStream = new JSONStream()
-    self.stream.pipe(this.jsonStream)
-    self.jsonStream.on('data', object => {
-      for (var i = 0; i < object.items.length; i++) {
-        self.emit('event', object.items[i])
-      }
+    // https://github.com/godaddy/kubernetes-client/blob/master/merging-with-kubernetes.md#getstream
+    self[namespace] = await this.client.api.v1.watch.namespaces(namespace).pods.getObjectStream()
+    self[namespace].on('data', object => {
+      // console.log(object)
+      self.emit('event', object)
     })
-    self.stream.on('error', function (err) {
+    self[namespace].on('error', function (err) {
       self.emit('error', err)
       self.stop()
-      self.startWatching()
+      self.startWatching(namespace, cb)
     })
+    cb()
   }
 
-  start () {
+  async startWatchingAllNamespaces (callback) {
+    const namespaces = await this.client.api.v1.namespaces.get()
+    for (var i = 0; i < namespaces.length; i++) {
+      this.startWatching(namespaces[i], callback)
```

Comment:
Is `startWatching` an `async` function, from line `25` in this file? Does it not need to be called with `await` here? Otherwise, they will be invoked in the `for` loop and execute in the event loop outside the main thread, while the sync code below the for loop continues on the main thread. Or am I missing something?

Reply:
Instead of await a callback is passed

Reply:
`startWatching` in line 25 is declared as async, because this function uses inside an `await` for k8s API.